src/listener.coffee

```diff
@@ -1,33 +1,79 @@
 {inspect} = require 'util'
+async     = require 'async'
 
 {TextMessage} = require './message'
 
 class Listener
   # Listeners receive every message from the chat source and decide if they
   # want to act on it.
+  # An identifier should be provided in the options parameter to uniquely
+  # identify the listener (options.id).
   #
   # robot    - A Robot instance.
   # matcher  - A Function that determines if this listener should trigger the
   #            callback.
+  # options  - An Object of additional parameters keyed on extension name
+  #            (optional).
   # callback - A Function that is triggered if the incoming message matches.
-  constructor: (@robot, @matcher, @callback) ->
+  constructor: (@robot, @matcher, @options, @callback) ->
+    if not @matcher?
+      throw new Error "Missing a matcher for Listener"
+
+    if not @callback?
+      @callback = @options
+      @options = {}
+
+    if not @options.id?
+      @options.id = null
+
+    if not @callback? or typeof @callback != 'function'
+      throw new Error "Missing a callback for Listener"
 
   # Public: Determines if the listener likes the content of the message. If
-  # so, a Response built from the given Message is passed to the Listener
-  # callback.
+  # so, a Response built from the given Message is passed through all
+  # registered middleware and potentially the Listener callback. Note that
+  # middleware can intercept the message and prevent the callback from ever
+  # being executed.
   #
   # message - A Message instance.
   #
-  # Returns a boolean of whether the matcher matched.
-  call: (message) ->
+  # @callback - Call with a boolean of whether the matcher matched.
+  #
+  # Returns nothing
+  # Returns before executing callback
+  call: (message, cb) ->
```

Comment:
Whoops. I'll update to not break the old behavior. The only breakage that *could* happen is if a caller was expecting the listener callback to begin execution before this function returns. Given that this is both an edge case and timing was never clearly specified, I think we are okay there.

I think this is probably marked as public because it is public from the class's perspective (called by Robot). We should probably look at having Hubot public APIs vs shared internal APIs, or something like that...